### PR TITLE
Add support for bitbucket cloud

### PR DESCRIPTION
**Pull request by eranturgeman**
- [ ] All [tests](https://github.com/jfrog/froggit-go/actions/workflows/test.yml) passed. If this feature is not already covered by the tests, I added new tests.
- [ ] I used `go fmt ./...` for formatting the code before submitting the pull request.
- [ ] This feature is included on all supported VCS providers - GitHub, Bitbucket cloud, Bitbucket server, GitLab and Azure Repos.
- [ ] I added the relevant documentation for the new feature.

---

This PR adds support for bitbucket cloud.
Enabling Auth though username (email) and password (token), as well as only Token.
Note: App password are deprecated in Atlassian, therefore not supported anymore but may be still work with existing passwords using the username + password auth.

Due to an existing bug in Bitbucket Cloud where archive download is rejected when using Bearer token, we introduced a workaround to clone using git clone (until the bug is fixed)

### CODE REVIEW

File: vcsclient/bitbucketcloud.go [hunk 1]
```diff
@@ -283,9 +309,12 @@ func (client *BitbucketCloudClient) GetCommitStatuses(ctx context.Context, owner
 	return results, err
 }
 
-// DownloadRepository on Bitbucket cloud
-func (client *BitbucketCloudClient) DownloadRepository(ctx context.Context, owner, repository, branch,
-	localPath string) error {
+func (client *BitbucketCloudClient) DownloadRepository(ctx context.Context, owner, repository, branch, localPath string) error {
+	// TODO: Once Atlassian fixes BCLOUD-23783, Bearer tokens will work for archive downloads.
```
Comment (eranturgeman) [inline vcsclient/bitbucketcloud.go]: When frogbot needs to scan a repository, it downloads the full source code. Bitbucket Cloud has a REST API endpoint that serves the repo as a zip archive — this is what DownloadRepository in          
  froggit-go uses normally. It's faster than cloning because it's just a single HTTP request that returns a zip.
                                                                                                                                                                                                          
  What is the BCLOUD-23783 issue?                                 

  Bitbucket Cloud's archive endpoint has a bug: it rejects Authorization: Bearer TOKEN and returns 401. This is an Atlassian-side bug, tracked as BCLOUD-23783, with no fix date. Since our no-username   
  flow uses Bearer for all REST API calls, hitting this endpoint returns 401 even though the token is perfectly valid — every other REST endpoint works fine with the same token.
                                                                                                                                                                                                          
  How the workaround fixes it?                                    

  When Username == "" (Bearer auth mode), DownloadRepository skips the archive endpoint entirely and instead does a git clone using:                                                                      
  https://x-token-auth:TOKEN@bitbucket.org/owner/repo.git
  This uses the git protocol over HTTPS with Basic Auth, which the git server accepts. After the clone:                                                                                                   
  1. The .git folder is stripped (it contains the URL with the embedded token)                                                                                                                            
  2. Files are copied to the target localPath                                                                                                                                                             
  3. CreateDotGitFolderWithRemote creates a clean .git with just the plain HTTPS URL (no credentials embedded)

File: vcsclient/bitbucketcloud.go [hunk 2]
```diff
@@ -1026,3 +1168,91 @@ func splitBitbucketCloudRepoName(name string) (string, string) {
 	}
 	return split[0], split[1]
 }
+
+// downloadRepositoryViaGitClone clones the repository using git with x-token-auth authentication.
+// This is a workaround for BCLOUD-23783: Bitbucket Cloud archive downloads do not support Bearer
+// token auth. Remove this fallback once Atlassian resolves BCLOUD-23783.
+func (client *BitbucketCloudClient) downloadRepositoryViaGitClone(ctx context.Context, owner, repository, branch, localPath string) (err error) {
+	client.logger.Debug("Using git clone fallback (BCLOUD-23783 workaround: Bearer tokens not supported for archive downloads)")
+	cloneURL := fmt.Sprintf("https://x-token-auth:%s@bitbucket.org/%s/%s.git", client.vcsInfo.Token, owner, repository)
```
Comment (attiasas) [inline vcsclient/bitbucketcloud.go]: > 🔴 🔒 Security

The fallback clone path constructs `https://x-token-auth:<token>@...` and includes git stderr in the returned error. This can leak secrets via process listings, logs, or propagated error messages.
Please consider switching to a credential-safe approach (e.g., credential helper/env-based auth) and sanitize all returned errors to ensure tokens are never exposed.

File: vcsclient/bitbucketcloud.go [hunk 3]
```diff
@@ -1026,3 +1168,91 @@ func splitBitbucketCloudRepoName(name string) (string, string) {
...
+	client.logger.Info(vcsutils.SuccessfulRepoExtraction)
+
+	repositoryInfo, err := client.GetRepositoryInfo(ctx, owner, repository)
+	if err != nil {
+		return err
+	}
+	return vcsutils.CreateDotGitFolderWithRemote(localPath, "origin", repositoryInfo.CloneInfo.HTTP)
+}
+
+func copyDir(src, dst string) error {
```
Comment (attiasas) [inline vcsclient/bitbucketcloud.go]: No need to reimplement logic, you can use `biutils.CopyDir()` (import `biutils "github.com/jfrog/build-info-go/utils"`)

File: vcsclient/bitbucketcloud.go [hunk 4]
```diff
@@ -1026,3 +1168,91 @@ func splitBitbucketCloudRepoName(name string) (string, string) {
...
+		}
+		dstPath := filepath.Join(dst, relPath)
+		if info.IsDir() {
+			return os.MkdirAll(dstPath, info.Mode())
+		}
+		return copyFile(path, dstPath, info.Mode())
+	})
+}
+
+func copyFile(src, dst string, mode os.FileMode) (err error) {
```
Comment (attiasas) [inline vcsclient/bitbucketcloud.go]: No need to reimplement logic; you can use biutils.CopyFile() (import biutils "github.com/jfrog/build-info-go/utils")

**Review by attiasas — COMMENTED** [review]
Nice work!, take a look at my comments.
In addition:

```
> 🟡 🧫 Testing

The new token-only download path and credential/error-sanitization behavior are not directly tested.
Add tests for username-empty `DownloadRepository` flow (including failure path) and assert that returned errors never contain token material.
```

**Comment by eranturgeman** [comment]
> Nice work!, take a look at my comments. In addition:
> 
> ```
> > 🟡 🧫 Testing
> 
> The new token-only download path and credential/error-sanitization behavior are not directly tested.
> Add tests for username-empty `DownloadRepository` flow (including failure path) and assert that returned errors never contain token material.
> ```

All fixed! I changed the way the construct the URL and added another protective mechanism for the stderr logs @attiasas

**Review by attiasas — APPROVED** [review]
LGTM
